### DISCUSSION TITLE
"cuBLAS.and.cuDNN" version differences?

### DISCUSSION THREAD

**Discussion by ghost** [Purfview/whisper-standalone-win#152] [Q&A] [answered]
What are differences between v2 and v3 Cublas and cudnn except kepler chip support, I am running it on RTX 3050. Also do I have to download the pytorch that in the repos even tho I installed pytorch 2.1.2 with pip? I asked it cus maybe its some fine tuned stuff

**Comment by Purfview** [comment, marked as answer]
> What are differences between v2 and v3 Cublas and cudnn except kepler chip support, I am running it on RTX 3050.

v3 currently contains the latest libs, v2 contains the last with Kepler support. I don't know the other differences, you can read the docs at NVIDIA site.
Or you can do own benchmarks, use longer audio and run 3 times with one libs then run 3 times with other libs, look at the transcription speeds. Then report the results of the benchmarks.

Use `-fallback=None` to ensure that benchmarks are consistent.

> Also do I have to download the pytorch that in the repos even tho I installed pytorch 2.1.2 with pip?

You don't need installed Python at all.

**Comment by Purfview** [comment]
Benchmarks posted below were reported by @despairTK :

Benchmark results of 1:15:42 long audio on `RTX 4090` with `--model=large-v2 --compute_type=float16 -fallback=None` 

libs v1:
![1](https://github.com/Purfview/whisper-standalone-win/assets/111871110/548c2a80-751b-4d8e-95b5-bfe7a1f93cc4)

libs v2:
![2](https://github.com/Purfview/whisper-standalone-win/assets/111871110/f50970fe-433b-4c61-987f-1573b1d3ec69)

libs v3:
![3](https://github.com/Purfview/whisper-standalone-win/assets/111871110/607b9c31-7e4b-4c10-9116-370882132177)

**Reply by ghost** [reply]
I was doing a Benchmark with rtx 3050 37 min audio `--model=large-v2 --sentence -fallback=None`

```
v3 
1 Transcription speed: 4.31 audio seconds/s
2 Transcription speed: 4.37 audio seconds/s
3 Transcription speed: 4.43 audio seconds/s
```
I am on v2 rn, but what is weird to me is that whisper-faster says that it runs on cuda but my gpu utilization is around %3-5 it feels like it does not use my gpu :D I do have CUDA 11.8

**Reply by Purfview** [reply]
If it says that it runs CUDA then you can be sure that it runs on GPU.
You can use `medium` model for faster benchmarks.

Btw, talking about libs, CTranslate2 is distributed with one cuDNN lib - `cudnn64_8`, I think it's v`8.1.1`, which is pretty old.
I wonder if it would make a difference to update it. I can make a test build with it updated.

**Reply by ghost** [reply]
Okay benchmark ended finally :D
Environment: Windows 10 Pro 22H2, Ryzen 5 5500 CPU, RTX 3050 GPU, 8x2 3200mhz DDR4 ram
Whisper command: --model=large-v2 --sentence -fallback=None
Audio: 37:48 min, opus, 48.000khz

V3 Results:
```
1st: Transcription speed: 4.31 audio seconds/s
2nd: Transcription speed: 4.37 audio seconds/s
3rd: Transcription speed: 4.43 audio seconds/s
Avg. speed: 4.37 audio seconds/s
```
V2 Results:
```
1st: Transcription speed: 4.52 audio seconds/s
2nd: Transcription speed: 4.21 audio seconds/s
3rd: Transcription speed: 4.21 audio seconds/s
Avg. speed: 4.31333333333 audio second/a
```

**Reply by Purfview** [reply]
Looks like v3 is insignificantly faster, practically equal. 
Benchmark `r167.2-test` vs "V3 Results", it's posted below.

Btw did you used the latest `r167.2` version?

**Reply by ghost** [reply]
I used non-test version but yeah latest one that i could find in the repo, started to benchmark rn with large-v2 again, i will send the benchmark probably 1-2 hours later cus I have to do something.

**Comment by Purfview** [comment]
Here you can download  `r167.2-test` for benchmarks -> https://we.tl/t-qZHfGTUxgc

It has `cudnn64_8` updated to v`8.9.6.50`, same version as the rest of cuDNN libs from the v3 libs pack.

**Reply by ghost** [reply]
v2
1st: Transcription speed: 4.32 audio seconds/s
2nd: Transcription speed: 4.23 audio seconds/s
3rd: Transcription speed: 4.33 audio seconds/s
Avg Speed: 4.29333333333
v3
1st: Transcription speed: 4.46 audio seconds/s
2nd: Transcription speed: 4.18 audio seconds/s
3rd: Transcription speed: 4.24 audio seconds/s
Avg Speed: 4.29333333333

**Reply by Purfview** [reply]
Looks like it doesn't influence anything, so we leave that `cudnn64_8` as is.

**Reply by Purfview** [reply]
Btw, v2 vs v3 libs difference could be in the different quantizations aka `--compute_type`.

**Reply by ghost** [reply]
too lazy to test them :D

**Reply by Purfview** [reply]
The full benchmarks should include all compute types. 😆
For an example, below are benchmarks between the different compute types on `GeForce RTX 4090`:

```
bfloat16：     Transcription speed: 27.99 audio seconds/s
float16：      Transcription speed: 27.46 audio seconds/s
int8_bfloat16：Transcription speed: 26.90 audio seconds/s
int8_float16： Transcription speed: 25.48 audio seconds/s
int8_float32： Transcription speed: 24.00 audio seconds/s
float32：      Transcription speed: 19.87 audio seconds/s

**Reply by ghost** [reply]
Next time i will do it like that 😆😆

**Reply by Purfview** [reply]
Use `medium` model for the faster tests.

**Reply by Purfview** [reply]
In your tests you didn't wrote what exact compute type did you tested.

And I just now noticed that there  are newer cuDNN libs released, there I uploaded them -> https://we.tl/t-ibgWViCW0b

**Reply by ghost** [reply]
@Purfview cus i use it on auto mode, i did nto see the cudnn libs so i did benchmark with old libs i am now int8_float32 of v3, it will be finished like in 30 min ig.

**Reply by Purfview** [reply]
> i use it on auto mode

You can see it with `--verbose=true`.

**Comment by ghost** [comment]
**Environment**: Windows 10 Pro 22H2, Ryzen 5 5500 CPU, RTX 3050 GPU, 8x2 3200mhz DDR4 ram
**Whisper command**: --model=medium --sentence -fallback=None
**Audio**: 37:48 min, opus, 48.000khz
**Whisper-Faster Version**: r167.2 non-test version.

**v2 Results:**
```
bfloat16:      Average Speed: 7.98 audio seconds/s
float16:       Average Speed: 7.46 audio seconds/s
int8_bfloat16: Average Speed: 6.18 audio seconds/s
int8_float16:  Average Speed: 5.88 audio seconds/s
float32:       Average Speed: 5.26 audio seconds/s
int8_float32:  Average Speed: 4.96 audio seconds/s
```
**v3 Results:**
```
bfloat16:      Average Speed: 8.21 audio seconds/s
float16:       Average Speed: 7.03 audio seconds/s
int8_float16:  Average Speed: 5.39 audio seconds/s
int8_bfloat16: Average Speed: 5.26 audio seconds/s
int8_float32:  Average Speed: 5.04 audio seconds/s
float32:       Average Speed: 4.87 audio seconds/s
```

**Reply by ghost** [reply]
The drops in v3 may be due to driver change, I downgraded from 546.33 to 537.58, if there are friends who are RTX users, they can test with both drivers and send the result.

**Reply by Purfview** [reply]
>The drops in v3 may be due to driver change, I downgraded from 546.33 to 537.58.

The top speed is still in v3.
Will you be testing those newer cuDNN libs?

**Reply by ghost** [reply]
alr edited my commenta as ur edited form, I will be testing them later in the today newer cudnn

**Reply by Purfview** [reply]
Btw, maybe you are eager to test CUDA v12? 😆

**Reply by ghost** [reply]
HAHAHAH alright :D I will

**Reply by Purfview** [reply]
OK, then I'll check if it's possible to make one.

**Reply by Purfview** [reply]
Btw, did you check with verbose what compute type is selected by auto?

**Reply by ghost** [reply]
I forgot it let me check

**Reply by ghost** [reply]
`[2023-12-21 20:02:43.208] [ctranslate2] [thread 18868] [info]  - Selected compute type: int8_float16` I still don't know what to select which compute type i should use since accuracy and stuff u know

**Reply by Purfview** [reply]
Expected... Allegedly `auto` should select the fastest. But it doesn't...
Maybe I need to write my own "auto".

>  I still don't know what to select which compute type i should use since accuracy and stuff u know

There is no "accuratest", so, imho don't bother.
On the different audios - the different quantizations can be more/less accurate.

**Reply by ghost** [reply]
alr so i will be using fastest one which is bfloat16 on my end

**Reply by Purfview** [reply]
In theory int8 should be a bit less accurate, in practice - I've seen the clear opposite effect on some audios.

**Reply by ghost** [reply]
Environment, commands etc. everything is the same as above, but since v3 is faster, I benchmarked only with v3, each `compute_type` was tested 3 times, but I only put the average values.

v3 [updated with v8.9.7.29] Results:

```
bfloat16:      Average Speed: 8.46 audio seconds/s
float16:       Average Speed: 7.27 audio seconds/s
int8_float16:  Average Speed: 6.79 audio seconds/s
int8_bfloat16: Average Speed: 6.59 audio seconds/s
int8_float32:  Average Speed: 5.26 audio seconds/s
float32:       Average Speed: 5.11 audio seconds/s
```

**Reply by Purfview** [reply]
Nice, looks like those new libs are a bit faster. `bfloat16` is faster by 3%.

**Comment by Purfview** [comment]
Here you go, test2 build with CUDA12 libs ->https://we.tl/t-H37FJfdgkF

**Reply by ghost** [reply]
Running CUDA 12 version with CUDA 11.8
```
bfloat16:      Average Speed: 7.82 audio seconds/s
float16:       Average Speed: 7.28 audio seconds/s
int8_bfloat16: Average Speed: 5.96 audio seconds/s
```
Will install CUDA 12.3 and try it again maybe 12.3 will work better on this.
Ahhhh CUDA 12.3 Setup fucked my computer.

**Reply by Purfview** [reply]
> Running CUDA 12 version with CUDA 11.8

Hmm, a sad drop in performance:
```
float16:      -10% drop in speed 
bfloat16:      -8% drop in speed
int8_bfloat16:  0% same
```

**Reply by ghost** [reply]
Then you should give us new release with best one :D

**Reply by Purfview** [reply]
Those are the latest libs, it's up for NVIDIA... 

> Ahhhh CUDA 12.3 Setup fucked my computer.

What happened?

**Reply by ghost** [reply]
I remember when I resetted my pc to windows 11, with large-v2 its speed around 8sec~ and it has nice gpu utilization around %90-100 I think we should identify this problem first
> Those are the latest libs, it's up for NVIDIA...

Did you update the repo for new version of libs? talking about cudn.
>What happened?

It updated my GPU driver its own and after then it, I got so many stuttering issues.

**Reply by Purfview** [reply]
> I remember when I resetted my pc to windows 11, with large-v2 its speed around 8sec~ and it has nice gpu utilization around %90-100 I think we should identify this problem first

Sorry, I don't understand what you meant.

> Did you update the repo for new version of libs? talking about cudn.

Not sure what you meant here too... that test2 archive includes the latest cuDNN for CUDA v12.

> It updated my GPU driver its own and after then it, I got so many stuttering issues.

Do you have benchmarks?

**Reply by ghost** [reply]
>Sorry, I don't understand what you meant.


Let me put it this way, when I bought my new computer, I installed Windows 11 and in that system, I was getting 4-45x audio second/s with whisper-faster and large-v2 model, then after resetting with Reset This PC, which is the built-in feature of Windows, when I used whisper-faster with large-v2, the audio second/s value was around 8 and when I looked in task manager, whisper-faster was really using the gpu, utilization was going between 90 and 100. But now I can't see those values both in terms of audio second/s and utilization.
>Not sure what you meant here too... that test2 archive includes the latest cuDNN for CUDA v12.

No, I'm talking about v3 [updated with v8.9.7.29] Doesn't this version use CUDA 11* I'm sure it was using CUDA 11, we switched to cuda12 after I benchmarked it.
>Do you have benchmarks?

No, I uninstalled it with DDU as soon as I noticed it.

**Reply by Purfview** [reply]
> No, I'm talking about v3 [updated with v8.9.7.29] Doesn't this version use CUDA 11* I'm sure it was using CUDA 11, we switched to cuda12 after I benchmarked it.

Now I'm even more confused... 😆
How this can be related to the test2 build tests?

**Reply by ghost** [reply]
It has no relation, don't you think it has best benchmarks so i think updating the libs repo to v8.9.7.29 would be nice

**Reply by Purfview** [reply]
Oh that, then commend under the right post... 😆
Yeah, maybe latter I'll upload "v4" for CUDA v11... and will delete that v3 as obsolete.

**Reply by Purfview** [reply]
If  CUDA 12 build of Faster-Whisper runs on CUDA 11.8, then I guess CUDA 11 build should run on CUDA 12 installs too.
So there is no point in updating Faster-Whisper with CUDA 12 and have a worse performance.

**Reply by ghost** [reply]
Yeah, I am just talking about "v4" stuff, there is no point update it to cuda12

**Reply by Purfview** [reply]
I wonder if you delete libs in test2-cuda12 and replace them with v3 [CUDA 11], would Faster-Whisper work?

**Reply by ghost** [reply]
so just delete all dlls and replace them with v3? I deleted the cuda12 version of whisper and weirdly i can't reach wetransfer rn, it stuck, i would give try.

**Reply by Purfview** [reply]
Yeah.

**Reply by ghost** [reply]
I don't have v3 rn, I have "v4" :D if u want i can try that.

**Reply by Purfview** [reply]
That works too, as long as CUDA 11 libs. But you still need "cuda12 version" build.

**Reply by Purfview** [reply]
Tested, wetransfer link works here.

**Reply by ghost** [reply]
No it does not work. (cuda12 version does not work with cuda11.8 installation.)
```
Standalone Faster-Whisper r167.2-test2-cuda12 running on: CUDA



Detecting language using up to the first 30 seconds. Use `--language` to specify the language
Traceback (most recent call last):
  File "D:\whisper-fast\__main__.py", line 1100, in <module>
  File "D:\whisper-fast\__main__.py", line 997, in cli
  File "faster_whisper\transcribe.py", line 340, in transcribe
  File "faster_whisper\transcribe.py", line 814, in encode
RuntimeError: Library cublas64_12.dll is not found or cannot be loaded
[6472] Failed to execute script '__main__' due to unhandled exception!
```

**Reply by Purfview** [reply]
Oh, another point not to update CTranslate2 to CUDA v12... 😆

I had big hopes with CUDA12...  [For you guys, as I don't use GPU with CUDA]

**Reply by ghost** [reply]
CUDA12 disappointed me as well :D

**Reply by ghost** [reply]
hehehe utilization problem solved, I enabled hardware-accelerated gpu scheduling.
same audio file (even the audio file is tired of being processed :D) --model=large-v2 --sentence -pp --compute_type=bfloat16 cuda 11.8 with v4
Transcription speed: 7.58 audio seconds/s

**Reply by ghost** [reply]
Give me the instructions, which dll which whisper-faster i can test :D

**Reply by Purfview** [reply]
> hehehe utilization problem solved,

But, it's slower now -> `7.58 audio seconds/s`, it was `7.82 audio seconds/s`.

**Reply by ghost** [reply]
No they were medium model not large-v2

**Reply by Purfview** [reply]
> No they were medium model not large-v2

Post what exact magic you need to do to get this magic improvement. Maybe as a new thread in Discussions.

> Give me the instructions, which dll which whisper-faster i can test :D

I've nothing new to test.

**Reply by ghost** [reply]
>Post what exact magic you need to do to get this magic improvement. Maybe as a new thread in Discussions.

Changed GPU's power management mode to max. performance. Enabled Hardware-accelerated GPU scheduling in window's setting.
And used https://github.com/hellzerg/optimizer and https://github.com/ChrisTitusTech/winutil, enabled optimize performance in optimizer's settings I dont remember what options did I use in chris's script.

**Reply by ghost** [reply]
tested cuda 12 one with v4 lib only twice, with large-v2.
Avg. 7.685 audio second/s

**Reply by Purfview** [reply]
I don't understand, "v4" are not cuda12 libs, no idea what you are testing there.

**Reply by ghost** [reply]
There were only 2 dlls on cuda12 one, they got no cublabs so thats why I tried with v4 and cudnn dll's name were same so yeah :D

**Reply by Purfview** [reply]
There are 4 dlls. Download it again.

**Reply by Purfview** [reply]
Did you tested "test2-cuda12" build with same "546.33"  driver, same as with "v3 [updated with v8.9.7.29]" aka  v4?

**Reply by Purfview** [reply]
Btw, when talking about the speed drop with cuda12 I got this comment:

> Check to disable nvidias "virtual" vram gpu feature they introduced in 12.x in windows which auto swap vram to host ram. Lots of users got caught with this killing perf.

**Reply by ghost** [reply]
Not yet, i will test them but i dont know when, prob. tomorrow, have some school work to do also could not find the setting that u mentionmed.

**Reply by Purfview** [reply]
Then what driver did you used there - > https://github.com/Purfview/whisper-standalone-win/discussions/152#discussioncomment-7923051 ?

**Reply by ghost** [reply]
546.33
![image](https://github.com/Purfview/whisper-standalone-win/assets/43301560/12d2a831-29ec-4093-b60e-b3509c8aa64a)
Is this the option u mentioned?

**Reply by Purfview** [reply]
They didn't specified, but from the description it looks like it.

**Reply by ghost** [reply]
I will do tests from scratch tomorrow everything is messed up rn lol with proper taken logs.

**Reply by Purfview** [reply]
Yeah. 😆

**Reply by ghost** [reply]
Btw, I forgot the mention, When I did not have the CUDA 12 in my computer my GPU Driver was 537.58, after installing the CUDA 12 it updated itself the 546.12, so I decided to update it to 546.33(studio version cus they do test too much on them) so, I will run these test from scratch but which driver version? and I will test them with float16 bfloat16 int8_bfloat16

**Reply by Purfview** [reply]
I dunno much about those versions, maybe the latest stable version.

Test `int8_float16` instead of int8_bfloat16.

**Reply by ghost** [reply]
I have to delete and reinstall gpu driver if I have to use CUDA 12 its requirement. so I tested with r167.2 cuda11.8 v2 v4, will install driver and cuda12. and test cuda12 version of faster whisper and cuda12version with v4.

**Reply by Purfview** [reply]
I dunno why you are constantly mentioning "v4", those libs  **are not for cuda12**.

**Reply by ghost** [reply]
aaahhh I read it wrong sorry :D

**Reply by Purfview** [reply]
Libs for cuda12 are inside the archive with `test2-cuda12` build.

**Reply by Purfview** [reply]
There is nothing related to CUDA12 at all in the Releases.
Everything there is CUDA11 only and is irrelevant to `test2-cuda12` build and tests with it.

**Reply by ghost** [reply]
Alright, I tested releases with CUDA 11.8 as you said, but I have to use 546.33 driver of gpu to test `test2-cuda12` because I use other ones forces to use cuda11 on my sys.

**Reply by ghost** [reply]
I am sleep deprived for 3-4 days have not sleep properly my brain is not functioning properly :D

**Reply by ghost** [reply]
**Environment**: Windows 10 Pro 22H2, Ryzen 5 5500 CPU, RTX 3050 GPU, 8x2 3200mhz DDR4 ram
**Whisper command**: --model=medium --sentence -fallback=None -pp --compute_type=
**Audio**: 37:48 min, opus, 48.000khz
**Whisper-Faster Version**: r167.2
**System's CUDA Version**: 11.8
**Nvidia Driver Version**: 537.58
each Models tested three times, in post results are avg. speeds.
I will test the cuda sysmem stuff in cuda12 it is not av. in cuda12. 

**v2 Results**:
```
bfloat16:      Average Speed: 13.52 audio seconds/s
float16:       Average Speed: 12.54 audio seconds/s
int8_float16:  Average Speed: 13.02 audio seconds/s

```
**v4 Results**:
```
bfloat16:      Average Speed: 13.54 audio seconds/s
float16:       Average Speed: 12.29 audio seconds/s
int8_float16:  Average Speed: 13.51 audio seconds/s

```
I think the real magic that separates these tests from the first v2 and v3 tests is the setting rather than the driver.
![image](https://github.com/Purfview/whisper-standalone-win/assets/43301560/a1b66324-37ca-4061-9e6c-8005eb107f3f)

**Reply by ghost** [reply]
Now I will install 546.33 so I can test the test2-cuda12

**Reply by ghost** [reply]
AAAAA Actually cuda12 may have backwards compatibility with 546.33, so I will also try cuda 11 versions of whisper-faster, I have a better idea :D My brain started working again

**Reply by Purfview** [reply]
I think it's irrelevant what CUDA is installed as long you have the right libs at the exe.

Now, `float16` is slower with v4, maybe some background program influenced the test result.

**Reply by ghost** [reply]
I'm a new win user dunno well, i could've change the env. to force use cuda12 without changing driver ig.

**Reply by ghost** [reply]
**Environment**: Windows 10 Pro 22H2, Ryzen 5 5500 CPU, RTX 3050 GPU, 8x2 3200mhz DDR4 ram
**Whisper command**: --model=medium --sentence -fallback=None -pp --compute_type=
**Audio**: 37:48 min, opus, 48.000khz
**Whisper-Faster Version**: test2-cuda12
**System's CUDA Version**: 12.3
**Nvidia Driver Version**: 546.33

**CUDA Sysmem Fallback Policy:** Driver Default
**Results**:
```
bfloat16:     Average Speed: 12.89 audio seconds/s
float16:      Average Speed: 12.08 audio seconds/s
int8_float16: Average Speed: 10.93 audio seconds/s
```
Welp i am bored i will benchmark sysmem fallback disabled in obscure time

**Reply by Purfview** [reply]
You need to disable "CUDA Sysmem Fallback",
And normal `r167.2` + v4 libs test on the same driver.

**Reply by ghost** [reply]
ok will do it in sometime

**Reply by ghost** [reply]
btw here is the new benchmark about cuda12 version.
**Environment**: Windows 10 Pro 22H2, Ryzen 5 5500 CPU, RTX 3050 GPU, 8x2 3200mhz DDR4 ram
**Whisper command**: --model=medium --sentence -fallback=None -pp --compute_type=
**Audio**: 37:48 min, opus, 48.000khz
**System's CUDA Version**: 12.3
**Nvidia Driver Version**: 546.33

**CUDA Sysmem Fallback Policy:** Pref. no sysmem fallback

**test2cudaw2Results**:
```
bfloat16:     Average Speed: 13.09  audio second/s
float16:      Average Speed: 12.17  audio second/s 
int8_float16: Average Speed: 10.80  audio second/s
```

**Reply by ghost** [reply]
**Environment**: Windows 10 Pro 22H2, Ryzen 5 5500 CPU, RTX 3050 GPU, 8x2 3200mhz DDR4 ram
**Whisper command**: --model=medium --sentence -fallback=None -pp --compute_type=
**Audio**: 37:48 min, opus, 48.000khz
**System's CUDA Version**: 12.3
**Nvidia Driver Version**: 546.33
**CUDA Sysmem Fallback Policy:** Pref. no sysmem fallback

**v4 Results:**
```
bfloat16:     Average Speed: 13.73 audio second/s
float16:      Average Speed: 12.24 audio second/s
int8_float16: Average Speed: 13.68 audio second/s

**Reply by Purfview** [reply]
In new tests sticks out that int8 is not slow anymore with CUDA11, it still slow with CUDA12.
Just that now `int8_float16` was tested, but I don't think that int8 types should differ much.
Maybe test `int8_bfloat16`,  `int8_float32` and `float32` too under this new environment.

Conclusion of the new tests:

```
float16:       -1% drop in speed 
bfloat16:      -5% drop in speed
int8_float16: -21% drop in speed
```

For the reference the old one [different environment]:

```
float16:      -10% drop in speed 
bfloat16:      -8% drop in speed
int8_bfloat16:  0% same
```

**Comment by ghost** [comment]
v4 for linux when @Purfview

**Reply by Purfview** [reply]
Did you switched to Linux?
"cuBLAS.and.cuDNN_linux_v4.7z" is uploaded now.

**Reply by ghost** [reply]
yup currently using arch
